net_ssl/ssl.pony

```diff
@@ -0,0 +1,257 @@
+    end
+
+    if server then
+      @SSL_set_accept_state[None](_ssl)
+    else
+      @SSL_set_connect_state[None](_ssl)
+      @SSL_do_handshake[I32](_ssl)
+    end
+  
+  fun box alpn_selected(): (ALPNProtocolName | None) =>
+    """
+    Get the protocol identifier negotiated via ALPN
+    """
+    var ptr: Pointer[U8] iso = recover Pointer[U8] end
+    var len = U32(0)
+    ifdef "openssl_1.1.x" then
+      @SSL_get0_alpn_selected[None](_ssl, addressof ptr, addressof len)
+    end
+
+    if ptr.is_null() then None
+    else
+      recover val String.copy_cpointer(consume ptr, USize.from[U32](len)) end
+    end
+
+  fun state(): SSLState =>
+    """
+    Returns the SSL session state.
+    """
+    _state
+
+  fun ref read(expect: USize = 0): (Array[U8] iso^ | None) =>
+    """
+    Returns unencrypted bytes to be passed to the application. If `expect` is
+    non-zero, the number of bytes returned will be exactly `expect`. If no data
+    (or less than `expect` bytes) is available, this returns None.
+    """
+    let offset = _read_buf.size()
+
+    var len = if expect > 0 then
+      if offset >= expect then
+        return _read_buf = recover Array[U8] end
+      end
+
+      expect - offset
+    else
+      1024
+    end
+
+    let max = if expect > 0 then expect - offset else USize.max_value() end
+    let pending = @SSL_pending[I32](_ssl).usize()
+
+    if pending > 0 then
+      if expect > 0 then
+        len = len.min(pending)
+      else
+        len = pending
+      end
+
+      _read_buf.undefined(offset + len)
+      @SSL_read[I32](_ssl, _read_buf.cpointer().usize() + offset, len.i32())
+    else
+      _read_buf.undefined(offset + len)
+      let r =
+        @SSL_read[I32](_ssl, _read_buf.cpointer().usize() + offset, len.i32())
+
+      if r <= 0 then
+        match @SSL_get_error[I32](_ssl, r)
+        | 1 | 5 | 6 => _state = SSLError
+        | 2 =>
+          // SSL buffer has more data but it is not yet decoded (or something)
+          _read_buf.truncate(offset)
+          return None
+        end
+
+        _read_buf.truncate(offset)
+      else
+        _read_buf.truncate(offset + r.usize())
+      end
+    end
+
+    let ready = if expect == 0 then
+      _read_buf.size() > 0
+    else
+      _read_buf.size() == expect
+    end
+
+    if ready then
+      _read_buf = recover Array[U8] end
+    else
+      // try and read again any pending data that SSL hasn't decoded yet
+      if @BIO_ctrl_pending[USize](_input) > 0 then
+        read(expect)
+      else
+        ifdef "openssl_1.1.x" then
+          // try and read again any data already decoded from SSL that hasn't
+          // been read via `SSL_has_pending` that was added in 1.1
+          // This mailing list post has a good description of what it is for:
+          // https://mta.openssl.org/pipermail/openssl-users/2017-January/005110.html
+          if @SSL_has_pending[I32](_ssl) == 1 then
+            read(expect)
+          else
+            None
```

Comment:
I think the `else None` blocks can be removed, since an if expression without an else block has a type of `(A | None)`